### PR TITLE
[Merged by Bors] - feat: port Data.Part

### PR DESCRIPTION
**Pull request by mcdoll**
ee0c179cd3c8a45aa5bffbf1b41d8dbede452865

### CODE REVIEW

**Comment by mcdoll** [comment]
Scott suggested using aesop instead, I can try to do that but not today. If you want to do it, feel free.

**Comment by ChrisHughes24** [comment]
I don't understand the CI error

**Comment by digama0** [comment]
The files have to be sorted. It's usually best to let the script update the file:
```bash
find Mathlib -name "*.lean" | env LC_ALL=C sort | sed 's/\.lean//;s,/,.,g;s/^/import /' > Mathlib.lean
```

**Comment by ChrisHughes24** [comment]
But they are in alphabetical order?

**Comment by digama0** [comment]
make that ASCIIbetical order. Whatever order is used by `LC_ALL=C sort`

File: Mathlib/Data/Part.lean [hunk 1]
```diff
@@ -0,0 +1,854 @@
...
+This file defines `Part α`, the partial values of a type.
+`o : Part α` carries a proposition `o.dom`, its domain, along with a function `get : o.dom → α`, its
+value. The rule is then that every partial value has a value but, to access it, you need to provide
+a proof of the domain.
+`Part α` behaves the same as `Option α` except that `o : Option α` is decidably `none` or `some a`
+for some `a : α`, while the domain of `o : Part α` doesn't have to be decidable. That means you can
+translate back and forth between a partial value with a decidable domain and an option, and
+`Option α` and `Part α` are classically equivalent. In general, `Part α` is bigger than `Option α`.
+In current mathlib, `Part ℕ`, aka `part_enat`, is used to move decidability of the order to
+decidability of `part_enat.find` (which is the smallest natural satisfying a predicate, or `∞` if
```
Comment (semorrison) [inline Mathlib/Data/Part.lean:20]:
```suggestion
decidability of `PartENat.find` (which is the smallest natural satisfying a predicate, or `∞` if
```

File: Mathlib/Data/Part.lean [hunk 2]
```diff
@@ -0,0 +1,854 @@
...
+#align part.bind_defined Part.bind_defined
+
+@[simp]
+theorem bind_dom {f : Part α} {g : α → Part β} : (f.bind g).Dom ↔ ∃ h : f.Dom, (g (f.get h)).Dom :=
+  Iff.rfl
+#align part.bind_dom Part.bind_dom
+
+section Instances
+
+-- We define several instances for constants and operations on `part α` inherited from `α`.
```
Comment (semorrison) [inline Mathlib/Data/Part.lean:658]:
```suggestion
-- We define several instances for constants and operations on `Part α` inherited from `α`.
```

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**